### PR TITLE
add cumsum prim backward

### PR DESCRIPTION
**Pull request by GGBond8488**
### PR types
Others 

### PR changes
 OPs 

### Describe
add cumsum prim backward

### CODE REVIEW

**Comment by GGBond8488** [comment]
由 https://github.com/PaddlePaddle/Paddle/pull/49518  引入的bug: 
cumsum_grad 在axis=None的场景下将产生与该pr合入之前不同的行为，且新的行为与torch以及autograd均不同，应该是错误的，所以这里没有测试axis=None的场景

**Review by JiabinYang — COMMENTED** [review]
add cinn test maybe

**Comment by GGBond8488** [comment]
> add cinn test maybe

see https://github.com/PaddlePaddle/CINN/blob/develop/cinn/frontend/op_mappers/paddle/cumsum.cc 

Compared with the paddle cumsum operator, CINN's cumsum operator lacks some parameters, and it cannot be aligned when connected to CINN. Consider adding operator mapping or enhancing the CINN cumsum operator in the future.

File: python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_cumsum_grad.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        self.x.stop_gradient = False
+        net = PrimeNet()
+        core._set_prim_backward_enabled(use_prim)
+        net = apply_to_static(net, use_cinn)
+        out = net(self.x)
+        res = paddle.autograd.grad(out, [self.x])
+
+        return res
+
+    # def test_cinn(self):
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/prim/prim/vjp/static/test_comp_cumsum_grad.py]: leave to do here and remove this test, no commented code should be pushed.

**Review by JiabinYang — COMMENTED** [review]
one comment

File: python/paddle/fluid/tests/unittests/test_cumsum_op.py [hunk 2]
```diff
@@ -115,20 +115,26 @@ def test_name(self):
...
+        self.prim_op_type = "prim"
+        self.python_api = paddle.cumsum
+        self.enable_cinn = False
         self.attrs = {'axis': 2}
         self.inputs = {'X': np.random.random((5, 6, 10)).astype("float64")}
         self.outputs = {'Out': self.inputs['X'].cumsum(axis=2)}
 
     def test_check_output(self):
-        self.check_output()
+        self.check_output(check_prim=True)
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_cumsum_op.py]: 基础算子没有前向拆分规则，不需要在前向测试中加上组合规则测试，可以去掉这个参数。

File: python/paddle/fluid/tests/unittests/test_cumsum_op.py [hunk 3]
```diff
@@ -138,35 +144,41 @@ def setUp(self):
         }
 
     def test_check_output(self):
-        self.check_output()
+        self.check_output(check_prim=True)
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_cumsum_op.py]: 同上

File: python/paddle/fluid/tests/unittests/test_cumsum_op.py [hunk 4]
```diff
@@ -138,35 +144,41 @@ def setUp(self):
...
+        self.prim_op_type = "prim"
+        self.python_api = paddle.cumsum
+        self.enable_cinn = False
         self.attrs = {'axis': 1}
         self.inputs = {'X': np.random.random((5, 6, 10)).astype("float64")}
         self.outputs = {'Out': self.inputs['X'].cumsum(axis=1)}
 
     def test_check_output(self):
-        self.check_output()
+        self.check_output(check_prim=True)
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_cumsum_op.py]: 同上

File: python/paddle/fluid/tests/unittests/test_cumsum_op.py [hunk 5]
```diff
@@ -138,35 +144,41 @@ def setUp(self):
...
+        self.prim_op_type = "prim"
+        self.python_api = paddle.cumsum
+        self.enable_cinn = False
         self.attrs = {'axis': 0}
         self.inputs = {'X': np.random.random((5, 6, 10)).astype("float64")}
         self.outputs = {'Out': self.inputs['X'].cumsum(axis=0)}
 
     def test_check_output(self):
-        self.check_output()
+        self.check_output(check_prim=True)
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_cumsum_op.py]: 同上

File: python/paddle/fluid/tests/unittests/test_cumsum_op.py [hunk 6]
```diff
@@ -175,11 +187,14 @@ def test_check_grad(self):
...
         self.op_type = "cumsum"
+        self.prim_op_type = "prim"
+        self.python_api = paddle.cumsum
+        self.enable_cinn = False
         self.inputs = {'X': np.random.random((5, 20)).astype("float64")}
         self.outputs = {'Out': self.inputs['X'].cumsum(axis=1)}
 
     def test_check_output(self):
-        self.check_output()
+        self.check_output(check_prim=True)
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_cumsum_op.py]: 同上

File: paddle/phi/api/yaml/op_compat.yaml [hunk 7]
```diff
@@ -341,6 +341,19 @@
   outputs :
     out : Out
 
+- op : cumsum
+  backward: cumsum_grad
+  inputs :
+    x : X
+  attrs :
+    axis : axis
```
Comment (zyfncg) [inline paddle/phi/api/yaml/op_compat.yaml]: name相同的属性这里的映射不用配置

Comment (GGBond8488) [reply paddle/phi/api/yaml/op_compat.yaml]: 已修改

File: python/paddle/fluid/tests/unittests/test_cumsum_op.py (hunk 2 above)
Comment (GGBond8488) [reply python/paddle/fluid/tests/unittests/test_cumsum_op.py]: 已去掉，下同

**Review by zyfncg — APPROVED** [review]
LGTM for op_compat

**Review by Charles-hit — APPROVED** [review]

File: paddle/fluid/operators/cum_op.cc [hunk 8]
```diff
@@ -100,6 +103,27 @@ class CumsumGradMaker : public framework::SingleGradOpMaker<T> {
...
+    paddle::experimental::Tensor x = this->GetSingleForwardInput("X");
+    paddle::experimental::Tensor out_grad = this->GetSingleOutputGrad("Out");
+    paddle::experimental::Tensor dx = this->GetSingleInputGrad("X");
+    auto* dx_ptr = this->GetOutputPtr(&dx);
+    std::string dx_name = this->GetOutputName(dx);
+    int axis = static_cast<int>(this->Attr<int>("axis"));
+    bool flatten = static_cast<bool>(this->Attr<bool>("flatten"));
+    bool exclusive = static_cast<bool>(this->Attr<bool>("exclusive"));
+    bool reverse = static_cast<bool>(this->Attr<bool>("reverse"));
+    VLOG(6) << "Runing add_grad composite func";
```
Comment (JiabinYang) [inline paddle/fluid/operators/cum_op.cc:120]: cumsum?

**Review by JiabinYang — COMMENTED** [review]
one comment

**Review by JiabinYang — APPROVED** [review]
LGTM, fix problem in next PR

**Pull request merged**